synapse/util/caches/lrucache.py

```diff
@@ -730,3 +730,31 @@ def __del__(self) -> None:
         # This happens e.g. in the sync code where we have an expiring cache of
         # lru caches.
         self.clear()
+
+
+class AsyncLruCache(Generic[KT, VT]):
+    """
+    An asynchronous wrapper around a subset of the LruCache API. On it's own
+    this doesn't change the behaviour but allows subclasses that utilize
+    external cache systems that require await behaviour to be created.
+    """
+
+    def __init__(self, *args, **kwargs):  # type: ignore
+        self.lru_cache: LruCache[KT, VT] = LruCache(*args, **kwargs)
+
+    async def get(
+        self, key: KT, default: Optional[T] = None, update_metrics: bool = True
+    ) -> Optional[VT]:
+        return self.lru_cache.get(key, update_metrics=update_metrics)
+
+    async def set(self, key: KT, value: VT) -> None:
+        self.lru_cache.set(key, value)
+
+    async def invalidate(self, key: KT) -> None:
+        return self.lru_cache.invalidate(key)
+
+    async def contains(self, key: KT) -> bool:
+        return self.lru_cache.contains(key)
+
+    def clear(self) -> None:
+        self.lru_cache.clear()
```

Comment:
This is only used in testing, is there a good way to enforce that?

Reply:
Not easily, but maybe we should just make it work properly?

Reply:
Yep! https://github.com/matrix-org/synapse/pull/13242/commits/8e0680309c6fa836df3c6688e9bd152718009de9